users.js

```diff
@@ -1185,34 +1186,37 @@ class User {
 		}
 		return true;
 	}
-	leaveRoom(room, connection, force) {
+	leaveRoom(room, connection, isDisconnecting = false) {
 		room = Rooms(room);
 		if (room.id === 'global') {
 			// you can't leave the global room except while disconnecting
-			if (!force) return false;
+			if (!isDisconnecting) return false;
 			this.cancelChallengeTo();
 			this.cancelSearch();
 		}
+
 		if (!this.inRooms.has(room.id)) {
 			return false;
 		}
-		for (let i = 0; i < this.connections.length; i++) {
-			if (connection && this.connections[i] !== connection) continue;
-			if (this.connections[i].inRooms.has(room.id)) {
-				this.connections[i].sendTo(room.id, '|deinit');
-				this.connections[i].leaveRoom(room);
-			}
-			if (connection) break;
-		}
 
-		let stillInRoom = false;
+		let connections = this.connections.filter(c => c.inRooms.has(room.id));
```

Comment:
This is a bit heavy code for users.js. What was wrong with the previous loop again?